File: LeetCode_Algorithm/412/412.py
from typing import List

class Solution:
    def fizzBuzz(self, n: int) -> List[str]:
        # sol1 O(n), 48~68ms 77~36%
        # ans = []
        # for i in range(1, n+1):
        #     string = ""
        #     if i % 3 == 0 and i % 5 == 0:
        #         string = "FizzBuzz"
        #     elif i % 5 == 0:
        #         string = "Buzz"
        #     elif i % 3 == 0:
        #         string = "Fizz"
        #     else:
        #         string = str(i)
            
        #     ans.append(string)
        # return ans


        # sol2 47~83ms 78~13%

        ans = []
        hashMap = {
            3 : "Fizz",
            5 : "Buzz"
        }
        
        for num in range(1, n+1):
            num_ans_str = ""

            for key in hashMap.keys():
                if (num % key) == 0:
                    num_ans_str += hashMap[key]
            
            if num_ans_str == "":
                num_ans_str = str(num)

            ans.append(num_ans_str)

        return ans

print(Solution().fizzBuzz(4))

File: LeetCode_Algorithm/412/412.js
var fizzBuzz = function(n) {
  // sol1 O(n) 58~74ms 99~81%
  // ans = []
  // for (let num=1; num < n+1; num++){
  //   numAnsStr = ""
  //   if (num <= 2){
  //     num = String(num)
  //     ans.push(num)
  //     continue 
  //   }

  //   if ((num % 3 == 0) && (num % 5 == 0)){
  //     numAnsStr = "FizzBuzz"
  //   }
  //   else if (num % 3 == 0){
  //     numAnsStr = "Fizz"
  //   }
  //   else if (num % 5 == 0 ){
  //     numAnsStr = "Buzz"
  //   }
  //   else {
  //     numAnsStr = String(num)
  //   }
  //   ans.push(numAnsStr)
  // }
  // return ans

  // sol2 95~114ms 53~27%보다 빠름
  ans = []
  hashMap = {
    3 : "Fizz",
    5 : "Buzz"
  }
  for (let num=1; num < n+1; num++){
    ansNumStr = ""
    for (let key in hashMap){
      if (num % key == 0){
        ansNumStr += hashMap[key]
      }
    }
    if (ansNumStr==""){
      ansNumStr = String(num)
    }
    ans.push(ansNumStr)
  }
  return ans

};

console.log(fizzBuzz(15))